pkg/recovery/etcd.go

```diff
@@ -0,0 +1,147 @@
+// The etcd backend fetches control plane objects directly from etcd. This is adapted heavily from
+// kubernetes/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go.
+
+package recovery
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/pkg/api"
+)
+
+// EtcdBackend is a backend that extracts a controlPlane from an etcd instance.
+type EtcdBackend struct {
+	client     *clientv3.Client
+	decoder    runtime.Decoder
+	pathPrefix string
+}
+
+// NewEtcdBackend constructs a new EtcdBackend for the given client and pathPrefix.
+func NewEtcdBackend(client *clientv3.Client, pathPrefix string) Backend {
+	return &EtcdBackend{
+		client:     client,
+		decoder:    api.Codecs.UniversalDecoder(),
+		pathPrefix: pathPrefix,
+	}
+}
+
+// read implements Backend.read().
+func (s *EtcdBackend) read(ctx context.Context) (*controlPlane, error) {
+	cp := &controlPlane{}
+	for _, r := range []struct {
+		etcdName string
+		yamlName string
+		obj      runtime.Object
+	}{{
+		etcdName: "configmaps",
+		yamlName: "config-map",
+		obj:      &cp.configMaps,
+	}, {
+		etcdName: "daemonsets",
+		yamlName: "daemonset",
+		obj:      &cp.daemonSets,
+	}, {
+		etcdName: "deployments",
+		yamlName: "deployment",
+		obj:      &cp.deployments,
+	}, {
+		etcdName: "poddisruptionbudgets",
+		yamlName: "pod-disruption-budget",
+		obj:      &cp.podDisruptionBudgets,
+	}, {
+		etcdName: "secrets",
+		yamlName: "secret",
+		obj:      &cp.secrets,
+	}, {
+		etcdName: "services/specs",
+		yamlName: "service",
+		obj:      &cp.services,
+	}, {
+		etcdName: "serviceaccounts",
+		yamlName: "service-account",
+		obj:      &cp.serviceAccounts,
+	},
```

Comment:
I changed it to only extract the bootstrap manifests, and added a little bit of code in `bootkube start` to suppress a scary-sounding error if there are no self-hosted manifests to install.